### PR TITLE
Fix NPEs in ProcessCommonJSModules caused by multiple var declarations

### LINKED ISSUE #2450: INTERNAL COMPILER ERROR when module.exports refers to vars using combined var statement
INTERNAL COMPILER ERROR when module.exports refers to vars using combined var statement

Problem was noticed by @atroche when trying to use https://github.com/joyent/node-bcrypt-pbkdf with ClojureScript.

Here is a minimal case which defines two vars using combined var statement and exposes those in `module.exports`. If both vars have their own `var`  statements no error occurs.

**foozz/bar.js**
```js
var BCRYPT_BLOCKS = 8,
    BCRYPT_HASHSIZE = 32;

module.exports = {
      BLOCKS: BCRYPT_BLOCKS,
      HASHSIZE: BCRYPT_HASHSIZE,
};
```

```
❯ java -jar closure-compiler-v20170409.jar --js foozz/bar.js --process_common_js_modules
java.lang.RuntimeException: java.lang.RuntimeException: INTERNAL COMPILER ERROR.
Please report this problem.

null
  Node(NAME BCRYPT_HASHSIZE): foozz/bar.js:2:4
    BCRYPT_HASHSIZE = 32;
  Parent(VAR): foozz/bar.js:1:0
var BCRYPT_BLOCKS = 8,

	at com.google.javascript.jscomp.CompilerExecutor.runInCompilerThread(CompilerExecutor.java:126)
	at com.google.javascript.jscomp.Compiler.runInCompilerThread(Compiler.java:753)
	at com.google.javascript.jscomp.Compiler.compile(Compiler.java:723)
	at com.google.javascript.jscomp.Compiler.compile(Compiler.java:693)
	at com.google.javascript.jscomp.AbstractCommandLineRunner.doRun(AbstractCommandLineRunner.java:1080)
	at com.google.javascript.jscomp.AbstractCommandLineRunner.run(AbstractCommandLineRunner.java:492)
	at com.google.javascript.jscomp.CommandLineRunner.main(CommandLineRunner.java:1898)
Caused by: java.lang.RuntimeException: INTERNAL COMPILER ERROR.
Please report this problem.

null
  Node(NAME BCRYPT_HASHSIZE): foozz/bar.js:2:4
    BCRYPT_HASHSIZE = 32;
  Parent(VAR): foozz/bar.js:1:0
var BCRYPT_BLOCKS = 8,

	at com.google.javascript.rhino.Node.replaceWith(Node.java:867)
	at com.google.javascript.jscomp.ProcessCommonJSModules$RewriteModule.updateNameReference(ProcessCommonJSModules.java:1063)
	at com.google.javascript.jscomp.ProcessCommonJSModules$RewriteModule.maybeUpdateName(ProcessCommonJSModules.java:942)
	at com.google.javascript.jscomp.ProcessCommonJSModules$RewriteModule.visit(ProcessCommonJSModules.java:704)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:639)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:711)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:635)
	at com.google.javascript.jscomp.NodeTraversal.traverseChildren(NodeTraversal.java:711)
	at com.google.javascript.jscomp.NodeTraversal.traverseBranch(NodeTraversal.java:635)
	at com.google.javascript.jscomp.NodeTraversal.traverse(NodeTraversal.java:311)
	at com.google.javascript.jscomp.NodeTraversal.traverseEs6(NodeTraversal.java:582)
	at com.google.javascript.jscomp.ProcessCommonJSModules.process(ProcessCommonJSModules.java:128)
	at com.google.javascript.jscomp.Compiler.processAMDAndCommonJSModules(Compiler.java:1856)
	at com.google.javascript.jscomp.Compiler.parseInputs(Compiler.java:1564)
	at com.google.javascript.jscomp.Compiler.parse(Compiler.java:810)
	at com.google.javascript.jscomp.Compiler.compileInternal(Compiler.java:763)
	at com.google.javascript.jscomp.Compiler.access$000(Compiler.java:87)
	at com.google.javascript.jscomp.Compiler$2.call(Compiler.java:726)
	at com.google.javascript.jscomp.Compiler$2.call(Compiler.java:723)
	at com.google.javascript.jscomp.CompilerExecutor$2.call(CompilerExecutor.java:91)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	... 24 more
```

### PR DESCRIPTION
**Pull request by ChadKillingsworth**
Fixes #2450 and replaces #2509.

Splits apart multiple var declarations in CommonJS modules to prevent an NPE.

### CODE REVIEW

**Comment by ChadKillingsworth** [comment]
I found an issue with this - working on a fix.

**Comment by ChadKillingsworth** [comment]
Alright now it should be good to go.

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java [hunk 1]
```diff
@@ -830,12 +830,18 @@ private void visitExport(NodeTraversal t, Node export) {
...
-        Node var = IR.var(updatedExport, rValue.detach()).useSourceInfoFrom(root.getParent());
-        parent.getParent().replaceWith(var);
+          && root.getParent().isAssign()) {
+        if (root.getParent().getParent().isExprResult()) {
+          // Rewrite "module.exports = foo;" to "var moduleName = foo;"
+          Node parent = root.getParent();
+          Node var = IR.var(updatedExport, rValue.detach()).useSourceInfoFrom(root.getParent());
+          parent.getParent().replaceWith(var);
+        } else if (root.getNext() != null && root.getNext().isName() && rValueVar.isGlobal()) {
+          // This is a where a module export assignment is used in a complex expression.
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: Could you add an example to the comment to show what this case looks like before and after the change?
Is this right?

```java
// Before: `SOME_VALUE !== undefined && module.exports = SOME_VALUE`
// After: `SOME_VALUE !== undefined && module$name = SOME_VALUE`
```

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java [hunk 2]
```diff
@@ -1078,6 +1091,11 @@ private void updateNameReference(
             }
           } else if (newNameDeclaration != null) {
             // Variable is already defined. Convert this to an assignment.
+            if (!nameRef.hasChildren()) {
+              parent.detachFromParent();
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: Please add a comment explaining why we detach the node and return early at this point.

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java [hunk 3]
```diff
@@ -1382,5 +1400,27 @@ private void fixTypeNode(NodeTraversal t, Node typeNode) {
...
+      while (var.getSecondChild() != null) {
+        Node newVar;
+        Node nameToSplit = var.removeFirstChild();
+        switch (var.getToken()) {
+          default:
+          case VAR:
+            newVar = IR.var(nameToSplit);
+            break;
+          case LET:
+            newVar = IR.var(nameToSplit);
```
Comment (brad4d) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: Did you mean to create `let` and `const` declarations here?

File: test/com/google/javascript/jscomp/ProcessCommonJSModulesTest.java [hunk 4]
```diff
@@ -868,4 +867,42 @@ public void testIssue2510() {
...
+  public void testWebpackAmdPattern() {
+    testModules(
+        "test.js",
+        LINE_JOINER.join(
+            "var __WEBPACK_AMD_DEFINE_ARRAY__, __WEBPACK_AMD_DEFINE_RESULT__;",
+            "!(__WEBPACK_AMD_DEFINE_ARRAY__ = [__webpack_require__(1), __webpack_require__(2)],",
+            "      __WEBPACK_AMD_DEFINE_RESULT__ = function (b, c) {",
+            "          console.log(b, c.exportA, c.exportB);",
+            "      }.apply(exports, __WEBPACK_AMD_DEFINE_ARRAY__),",
+            "    __WEBPACK_AMD_DEFINE_RESULT__ !== undefined && (module.exports = __WEBPACK_AMD_DEFINE_RESULT__));"),
```
Comment (brad4d) [inline test/com/google/javascript/jscomp/ProcessCommonJSModulesTest.java]: line too long

File: test/com/google/javascript/jscomp/ProcessCommonJSModulesTest.java [hunk 5]
```diff
@@ -868,4 +867,42 @@ public void testIssue2510() {
...
+        LINE_JOINER.join(
+            "goog.provide('module$test');",
+            "var module$test = {};",
+            "var __WEBPACK_AMD_DEFINE_ARRAY__$$module$test;",
+            "!(__WEBPACK_AMD_DEFINE_ARRAY__$$module$test = [__webpack_require__(1), __webpack_require__(2)],",
+            "    module$test = function(b,c){console.log(b,c.exportA,c.exportB)}",
```
Comment (brad4d) [inline test/com/google/javascript/jscomp/ProcessCommonJSModulesTest.java]: line longer than 100 chars

**Comment by brad4d** [comment]
Imported and running through tests, but I also have some review comments.

Please don't squash when you push your fixes. Thanks.

File: src/com/google/javascript/jscomp/ProcessCommonJSModules.java (hunk 3 above)
Comment (ChadKillingsworth) [inline src/com/google/javascript/jscomp/ProcessCommonJSModules.java]: Sure did. This pass most certainly shouldn't be transpiling ...

**Comment by ChadKillingsworth** [comment]
Review issues fixed.

**Comment by brad4d** [comment]
Imported latest changes.
Test results look good.
Sending for internal review.
